results/pkg/api/server/server.go

```diff
@@ -155,54 +163,186 @@ func (s *Server) ListResultsResult(ctx context.Context, req *pb.ListResultsReque
-	}
-	var resp []*pb.Result
-	for _, r := range results {
-		for _, e := range r.Executions {
-			out, _, err := prg.Eval(map[string]interface{}{
-				"taskrun": e.GetTaskRun(),
-			})
-			if err != nil {
-				log.Printf("failed to evaluate the expression: %v", err)
-				return nil, status.Errorf(codes.InvalidArgument, "Error occurred during filter evaluation step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
+		return 0, err
+	}
+	p := &pb.Result{}
+	if err := proto.Unmarshal(result, p); err != nil {
+		log.Printf("unmarshaling error: %v", err)
+		return 0, status.Errorf(codes.Internal, "failed to parse result data: %v", err)
+	}
+	var primaryKey uint64
+	if err := s.db.QueryRow("SELECT id FROM results where name=?", p.GetName()).Scan(&primaryKey); err != nil {
+		return 0, err
+	}
+	return primaryKey, nil
+}
+
+// This function aims to obtain a fixed size `pageSize` of results from the database, starting
+// from the results with the primary key `startPK`, filtered by a compiled CEL program `prg`.
+//
+// In this function, we query the database multiple times and filter the queried results to
+// comprise the final results.
+//
+// To minimize the query times, we introduce a variable `ratio` to indicate the retention rate
+// after filtering a batch of results. The ratio of the queried batch is:
+//             ratio = remained_results_size/batch_size.
+//
+// The batchSize depends on the `ratio` of the previous batch and the `pageSize`:
+//                  batchSize = pageSize/last_ratio
+// The less the previous ratio is, the bigger the upcoming batch_size is. Then the queried time
+// is significantly decreased.
+func (s Server) getFilteredPaginatedResults(pageSize int32, startPK uint64, prg cel.Program) (results []*pb.Result, lastPK uint64, err error) {
+	if startPK == 0 {
+		lastPK = startPK
+	} else {
+		lastPK = startPK - 1
+	}
+	var ratio float32 = 1 // the retention rate after filtering a batch of results.
+	// If didn't get enought results.
+	for int32(len(results)) < pageSize {
+		var (
+			batchSize     int32 // batchSize = math.Ceil(pageSize/ratio), has a maximum.
+			batchGot      int32 // <= batchSize, the actual size of a batch query.
+			batchFiltered int32 // <= batchGot, the filtered size of a batch query.
+		)
+		if math.Ceil(float64(pageSize)/float64(ratio)) > float64(listResultsMaximumPageSize) {
+			batchSize = listResultsMaximumPageSize
+		} else {
+			batchSize = int32(math.Ceil(float64(pageSize) / float64(ratio)))
+		}
+		rows, err := s.db.Query("SELECT id, data from results where id > ? limit ? ", lastPK, batchSize)
```

Comment:
Something I realized from the other db query below- these should all happen in the same transaction.